File: ios/RecipeClipper/Data/Remote/WebViewRenderedPageSource.swift
import Foundation
import WebKit

/// RenderedPageSource over an off-screen WKWebView: never in a window, never shown (Android's
/// WebViewRenderedPageSource). Beside PathConnectivity because, like it, it is a platform
/// service kept behind a protocol, so the repository never imports WebKit's types.
///
/// Each call builds a fresh web view on the main actor, loads the page with JavaScript on,
/// waits `settle` after the last `didFinish` (a bot check or a script that navigates starts
/// the wait again), then reads `document.documentElement.outerHTML`. The web view is released
/// on every way out: success, failure or cancellation. The overall cap is the repository's,
/// which cancels this when it runs out.
///
/// The user agent is WebKit's own: this is a real browser engine, not a user-agent trick (see
/// CLAUDE.md, "Failure handling"). Its cookies are the app's, not Safari's, so a paywall still
/// fails. Not used inside the share extension (memory, #19).
@MainActor
final class WebViewRenderedPageSource: RenderedPageSource {
    /// How long a page must stay quiet after loading before its HTML is taken: time for scripts
    /// that add the recipe data, or finish a bot check, after the load event. Android's SETTLE_MS.
    static let settle: Duration = .milliseconds(1500)

    func render(url: String) async -> String? {
        // http(s) only, as the direct fetch enforces.
        guard let target = URL(string: url),
              let scheme = target.scheme?.lowercased(), scheme == "http" || scheme == "https"
        else { return nil }
        let loader = PageLoader()
        return await withTaskCancellationHandler {
            await loader.load(target)
        } onCancel: {
            Task { @MainActor in loader.cancel() }
        }
    }
}

/// One page load: owns the web view and the continuation, and resumes it exactly once.
@MainActor
private final class PageLoader: NSObject, WKNavigationDelegate {
    private var webView: WKWebView?
    private var continuation: CheckedContinuation<String?, Never>?
    private var pendingCapture: Task<Void, Never>?
    private var cancelled = false

    func load(_ url: URL) async -> String? {
        await withCheckedContinuation { continuation in
            guard !cancelled else { return continuation.resume(returning: nil) }
            self.continuation = continuation
            let configuration = WKWebViewConfiguration()
            configuration.mediaTypesRequiringUserActionForPlayback = .all
            // A phone-sized frame, so layout-dependent scripts see a phone viewport.
            let webView = WKWebView(frame: CGRect(x: 0, y: 0, width: 390, height: 844), configuration: configuration)
            webView.navigationDelegate = self
            self.webView = webView
            webView.load(URLRequest(url: url))
        }
    }

    func cancel() {
        cancelled = true
        finish(nil)
    }

    private func finish(_ html: String?) {
        pendingCapture?.cancel()
        pendingCapture = nil
        webView?.stopLoading()
        webView?.navigationDelegate = nil
        webView = nil
        continuation?.resume(returning: html)
        continuation = nil
    }

    private func capture() {
        guard let webView else { return }
        webView.evaluateJavaScript("document.documentElement.outerHTML") { [weak self] result, _ in
            // WebKit calls this on the main thread.
            MainActor.assumeIsolated { self?.finish(result as? String) }
        }
    }

    // MARK: WKNavigationDelegate

    func webView(_ webView: WKWebView, didStartProvisionalNavigation navigation: WKNavigation!) {
        // A new navigation (a redirect, or a bot check passing): wait for it instead.
        pendingCapture?.cancel()
    }

    func webView(_ webView: WKWebView, didFinish navigation: WKNavigation!) {
        pendingCapture?.cancel()
        pendingCapture = Task { [weak self] in
            try? await Task.sleep(for: WebViewRenderedPageSource.settle)
            guard !Task.isCancelled else { return }
            self?.capture()
        }
    }

    func webView(_ webView: WKWebView, didFail navigation: WKNavigation!, withError error: Error) {
        failed(error)
    }

    func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        failed(error)
    }

    private func failed(_ error: Error) {
        // A navigation cut short by the next one (a script redirect) isn't a failure: the next
        // one's didFinish takes over.
        if (error as NSError).domain == NSURLErrorDomain, (error as NSError).code == NSURLErrorCancelled { return }
        finish(nil)
    }

    func webViewWebContentProcessDidTerminate(_ webView: WKWebView) {
        finish(nil)
    }
}

File: app/src/main/java/com/example/recipeclipper/data/WebViewRenderedPageSource.kt
package com.example.recipeclipper.data

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.os.Handler
import android.os.Looper
import android.view.View
import android.webkit.RenderProcessGoneDetail
import android.webkit.WebView
import android.webkit.WebViewClient
import com.example.recipeclipper.data.remote.RenderedPageSource
import dagger.hilt.android.qualifiers.ApplicationContext
import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.suspendCancellableCoroutine
import kotlinx.coroutines.withContext
import org.json.JSONTokener
import javax.inject.Inject
import javax.inject.Singleton
import kotlin.coroutines.resume

/**
 * [RenderedPageSource] over an off-screen [WebView]: never attached to a window, never shown.
 * Beside [AndroidConnectivity] because, like it, it is a `Context` thing kept behind an
 * interface so the repository never sees one.
 *
 * Each call builds a fresh WebView on the main thread (WebView demands it), loads the page with
 * JavaScript on, waits [SETTLE_MS] after the last `onPageFinished` (a bot check or a script
 * that navigates starts the wait again), then reads `document.documentElement.outerHTML`. The
 * WebView is destroyed on every way out: success, failure or cancellation. The overall cap is
 * the repository's, so a page that never settles is cut off there and cleaned up here.
 *
 * The WebView's user agent is left as the system's own: this is a real browser engine, not a
 * user-agent trick (see CLAUDE.md, "Failure handling"). Its cookies are the app's, not the
 * user's browser's, so a paywall still fails.
 */
@Singleton
class WebViewRenderedPageSource @Inject constructor(
    @ApplicationContext private val context: Context
) : RenderedPageSource {

    override suspend fun render(url: String): String? = withContext(Dispatchers.Main) {
        val webView = try {
            WebView(context)
        } catch (e: Exception) {
            // No WebView provider (being updated, or disabled): nothing to render with.
            return@withContext null
        }
        try {
            load(webView, url)
        } catch (e: CancellationException) {
            throw e
        } catch (e: Exception) {
            null
        } finally {
            // On the main thread still, as destroy() requires, whichever way we left.
            webView.stopLoading()
            webView.destroy()
        }
    }

    @SuppressLint("SetJavaScriptEnabled") // the point: pages whose recipe data needs scripts
    private suspend fun load(webView: WebView, url: String): String? =
        suspendCancellableCoroutine { continuation ->
            val handler = Handler(Looper.getMainLooper())
            var pendingCapture: Runnable? = null

            fun finish(html: String?) {
                pendingCapture?.let(handler::removeCallbacks)
                if (continuation.isActive) continuation.resume(html)
            }

            fun capture() {
                webView.evaluateJavascript("document.documentElement.outerHTML") { result ->
                    finish(decodeJsString(result))
                }
            }

            webView.settings.apply {
                javaScriptEnabled = true
                domStorageEnabled = true // many sites' scripts fail without localStorage
                blockNetworkImage = true // photos aren't needed to read the recipe data
            }
            // Laid out at the screen's size, so layout-dependent scripts see a phone viewport.
            val metrics = context.resources.displayMetrics
            webView.measure(
                View.MeasureSpec.makeMeasureSpec(metrics.widthPixels, View.MeasureSpec.EXACTLY),
                View.MeasureSpec.makeMeasureSpec(metrics.heightPixels, View.MeasureSpec.EXACTLY)
            )
            webView.layout(0, 0, metrics.widthPixels, metrics.heightPixels)

            webView.webViewClient = object : WebViewClient() {
                override fun onPageStarted(view: WebView, url: String?, favicon: Bitmap?) {
                    // A new navigation (a redirect, or a bot check passing): wait for it instead.
                    pendingCapture?.let(handler::removeCallbacks)
                }

                override fun onPageFinished(view: WebView, url: String?) {
                    pendingCapture?.let(handler::removeCallbacks)
                    pendingCapture = Runnable { capture() }.also { handler.postDelayed(it, SETTLE_MS) }
                }

                // No onReceivedError: a page that fails to load still ends in onPageFinished, on
                // the WebView's error page, which parses as no recipe and so leaves the direct
                // fetch's cause standing. Finishing early on an error would also end the load
                // when a script redirect aborts the first navigation.

                override fun onRenderProcessGone(view: WebView, detail: RenderProcessGoneDetail): Boolean {
                    // Returning true keeps a crashed renderer from taking the app down with it.
                    finish(null)
                    return true
                }
            }
            continuation.invokeOnCancellation { pendingCapture?.let(handler::removeCallbacks) }
            webView.loadUrl(url)
        }

    companion object {
        /** How long a page must stay quiet after loading before its HTML is taken: time for
         *  scripts that add the recipe data, or finish a bot check, after the load event. */
        const val SETTLE_MS = 1_500L

        /** `evaluateJavascript` hands back its result JSON-encoded: a quoted string, or `null`. */
        internal fun decodeJsString(json: String?): String? {
            if (json == null) return null
            return try {
                JSONTokener(json).nextValue() as? String
            } catch (e: Exception) {
                null
            }
        }
    }
}
